### PR TITLE
Api endpoints

### PR DESCRIPTION
**Pull request by Martinfree**
Comparing for continue working with next tasks.

Trello cards:

1. https://trello.com/c/LTYCqnrX/67-%D0%BE%D0%B1%D0%BD%D0%BE%D0%B2%D0%B8%D1%82%D0%B8-%D0%B0%D0%BF%D1%96-%D0%B5%D0%BD%D0%B4%D0%BF%D0%BE%D0%B9%D0%BD%D1%82%D0%B8-%D0%B4%D0%BB%D1%8F-%D0%B0%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D1%96%D1%97

2. https://trello.com/c/0tzTdWKC/66-%D0%BE%D0%B1%CA%BC%D1%94%D0%B4%D0%BD%D0%B0%D1%82%D0%B8-%D0%B0%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D1%96%D1%8E-allauth-jwt-%D1%82%D0%B0-%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D1%8C-%D0%BF%D0%BE%D1%82%D0%BE%D1%87%D0%BD%D0%BE%D0%B3%D0%BE-%D0%BA%D0%BE%D1%80%D0%B8%D1%81%D1%82%D1%83%D0%B2%D0%B0%D1%87%D0%B0 (allauth problems not resolved yet, but still authenticate)

3. https://trello.com/c/6ZBVgQWG/78-%D0%B4%D0%BE%D0%B4%D0%B0%D1%82%D0%B8-%D0%BC%D0%BE%D0%B4%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%83-%D0%BC%D0%BE%D0%B6%D0%BB%D0%B8%D0%B2%D1%96%D1%81%D1%82%D1%8C-%D0%B2%D1%96%D0%B4%D0%BF%D1%80%D0%B0%D0%B2%D0%BA%D0%B8-%D0%BB%D0%B8%D1%81%D1%82%D1%96%D0%B2-%D0%BD%D0%B0%D0%BF%D1%80%D1%8F%D0%BC%D1%83-%D0%BA%D0%BE%D1%80%D0%B8%D1%81%D1%82%D1%83%D0%B2%D0%B0%D1%87%D1%83

### CODE REVIEW

File: api/authentication/models.py [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
+from django.template.loader import render_to_string
 from django.core.mail import send_mail
 from django.db import models
 from django.utils import timezone
+from django.core.signing import TimestampSigner, b64_encode,b64_decode, BadSignature, SignatureExpired,force_bytes
+from django.core.mail import EmailMessage
 from django.contrib.auth.models import UserManager
-from django.contrib.auth.models import AbstractUser, Group
+from django.contrib.auth.models import AbstractBaseUser, AbstractUser, Group
+from settings import base
```
Comment (pavlo-mk) [inline api/authentication/models.py:12]: dont use this import in future its bad practice
please use example in [docs](https://docs.djangoproject.com/en/2.2/topics/settings/#using-settings-in-python-code) `from django.conf import settings`

File: api/authentication/models.py [hunk 2]
```diff
@@ -131,23 +164,105 @@ def get_short_name(self):
...
+                body=render_to_string('authentication/mail/verification_body.html',context),
+                to=[email])
+        msg.content_subtype = 'html'
+        msg.send()
+
+    def send_recovery_password(self, email):
+        verification_code = self.get_verification_code(email=email)
+        
+        context = {'user': self,
+                    'settings': base,
```
Comment (pavlo-mk) [inline api/authentication/models.py]: why you need this option `settings` in context template ??

File: api/authentication/models.py [hunk 3]
```diff
@@ -131,23 +164,105 @@ def get_short_name(self):
         short_name = "%s" % self.first_name
         return short_name.strip()
```
Comment (pavlo-mk) [inline api/authentication/models.py:166]: you can make simple code here like `return self.first_name.strip()` without variable

File: api/authentication/models.py [hunk 4]
```diff
@@ -131,23 +164,105 @@ def get_short_name(self):
...
         pass
 
+    def get_verification_code(self,email):
+        # verification token 
+        signer = TimestampSigner()
+        return b64_encode(bytes(signer.sign(email), encoding='utf-8'))
+        
+    @classmethod
+    def verify_email(self,code):
+        if code:
```
Comment (pavlo-mk) [inline api/authentication/models.py:179]: what will be if no code in args? please update logic for this method

File: api/authentication/models.py [hunk 5]
```diff
@@ -131,23 +164,105 @@ def get_short_name(self):
...
+                code = code.decode()
+                email = signer.unsign(code, max_age=max_age)
+                
+                user = StdUser.objects.get(**{StdUser.USERNAME_FIELD:email,'is_active':False})
+                user.is_active = True
+                user.save()
+                return True, ('Your account has been activated.')  
+            except SignatureExpired:
+                return False, ('Your time to activate by link expired')
+            except (BadSignature, StdUser.DoesNotExist, TypeError, UnicodeDecodeError) as e:
```
Comment (pavlo-mk) [inline api/authentication/models.py:199]: second `except` not need at this place, please check the logic

File: api/authentication/models.py [hunk 6]
```diff
@@ -131,23 +164,105 @@ def get_short_name(self):
...
+                print(e)
+                pass
+            return False, ('Activation link is incorrect, please resend request')
+            
+    def send_mail(self,email):
+        verification_code = self.get_verification_code(email=email)
+        #self.code = verification_code
+        #self.save()
+        context = {'user': self,
+                'settings': base,
```
Comment (pavlo-mk) [inline api/authentication/models.py]: the same question with `base` var, where you will use it?

File: api/authentication/serializers.py [hunk 7]
```diff
@@ -15,8 +23,148 @@ class Meta(object):
...
+    class Meta(object):
+        model = User
+        fields = (
+            'email',
+            'password',
+        )
+
+    def create(self, validated_data):
+
+        if validate_password(password=validated_data.get('password',), user=validated_data.get('email'), password_validators=None) is not None:
```
Comment (pavlo-mk) [inline api/authentication/serializers.py:39]: small changes: many chars in one line, please use in your practice 120 per line

File: api/authentication/serializers.py [hunk 8]
```diff
@@ -15,8 +23,148 @@ class Meta(object):
...
+    class Meta(object):
+        model = User
+        
+        fields = (
+                'email',
+                )
+
+    def post(self, data):
+         email = data.get('email')                                                          
+         user = User.objects.get(email=email).first()
```
Comment (pavlo-mk) [inline api/authentication/serializers.py:164]: `get()` method return only one object

File: api/authentication/views.py [hunk 9]
```diff
@@ -1,48 +1,201 @@
+import settings
```
Comment (pavlo-mk) [inline api/authentication/views.py:1]: the same dont use `import settings` its bad practice

**Review by pavlo-mk — CHANGES_REQUESTED** [review]
Added small comments for changes please let me know if you have a question.

**Comment by pavlo-mk** [comment]
@Martinfree please add description for PR, trello card link, description changes, new features.

File: api/authentication/models.py (hunk 2 above)
Comment (Martinfree) [reply api/authentication/models.py]: Because in html template i need url for redirect link, which in base.py

**Comment by Martinfree** [comment]
Trelli cards:
1. https://trello.com/c/LTYCqnrX/67-%D0%BE%D0%B1%D0%BD%D0%BE%D0%B2%D0%B8%D1%82%D0%B8-%D0%B0%D0%BF%D1%96-%D0%B5%D0%BD%D0%B4%D0%BF%D0%BE%D0%B9%D0%BD%D1%82%D0%B8-%D0%B4%D0%BB%D1%8F-%D0%B0%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D1%96%D1%97
2. https://trello.com/c/0tzTdWKC/66-%D0%BE%D0%B1%CA%BC%D1%94%D0%B4%D0%BD%D0%B0%D1%82%D0%B8-%D0%B0%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D1%96%D1%8E-allauth-jwt-%D1%82%D0%B0-%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D1%8C-%D0%BF%D0%BE%D1%82%D0%BE%D1%87%D0%BD%D0%BE%D0%B3%D0%BE-%D0%BA%D0%BE%D1%80%D0%B8%D1%81%D1%82%D1%83%D0%B2%D0%B0%D1%87%D0%B0 (allauth problems)

File: api/authentication/models.py (hunk 2 above)
Comment (pavlo-mk) [reply api/authentication/models.py]: You don't need export all settings file into html, its  not secure. You would add only one variable please use `settings.LOGIN_REDIRECT_URL` . <<its ONLY EXAMPLE

File: api/authentication/models.py (hunk 4 above)
Comment (Martinfree) [reply api/authentication/models.py:179]: If we don't have any code - you not to be here, that function calling once with create user

**Comment by pavlo-mk** [comment]
^^ please add this links in to main description of this PR. Edit from above please))

File: api/authentication/models.py (hunk 2 above)
Comment (Martinfree) [reply api/authentication/models.py]: How i verify user in LOGIN_REDIRECT_URL?

File: api/authentication/models.py (hunk 4 above)
Comment (pavlo-mk) [reply api/authentication/models.py:179]: What happens when this if statement fails?

File: api/authentication/models.py (hunk 2 above)
Comment (pavlo-mk) [reply api/authentication/models.py]: its only example above. You can't import all settings in context, If you have any secrets in settings, What happens? Is it secure?

Comment (Martinfree) [reply api/authentication/models.py]: Ok, i'll take it. Thanks for the advice, soon i'll change it

**Pull request closed without merging**